src/lazy/expanded/mod.rs

```diff
@@ -232,45 +250,63 @@ impl<Encoding: Decoder, Input: IonInput> ExpandingReader<Encoding, Input> {
     pub fn register_template(&mut self, template_definition: &str) -> IonResult<MacroAddress> {
-        let context = self.context();
-        let template_macro: TemplateMacro =
-            { TemplateCompiler::compile_from_text(context.get_ref(), template_definition)? };
+        let template_macro: TemplateMacro = self.compile_template(template_definition)?;
+        self.add_macro(template_macro)
+    }
 
-        let macro_table = self.macro_table.get_mut();
+    fn compile_template(&self, template_definition: &str) -> IonResult<TemplateMacro> {
+        TemplateCompiler::compile_from_text(self.context(), template_definition)
+    }
+
+    fn add_macro(&mut self, template_macro: TemplateMacro) -> IonResult<MacroAddress> {
+        let macro_table = &mut self.context_mut().macro_table;
         macro_table.add_macro(template_macro)
     }
 
-    pub fn context(&self) -> EncodingContext<'_> {
+    pub fn context(&self) -> EncodingContextRef<'_> {
         // SAFETY: The only time that the macro table, symbol table, and allocator can be modified
         // is in the body of the method `between_top_level_expressions`. As long as nothing holds
         // a reference to the `EncodingContext` we create here when that method is running,
         // this is safe.
-        unsafe {
-            EncodingContext::new(
-                &*self.macro_table.get(),
-                &*self.symbol_table.get(),
-                &*self.allocator.get(),
-            )
-        }
+        unsafe { (*self.encoding_context.get()).get_ref() }
+    }
+
+    pub fn context_mut(&mut self) -> &mut EncodingContext {
+        // SAFETY: If the caller has a `&mut` reference to `self`, it is the only mutable reference
+        //         that can modify `self.encoding_context`.
+        unsafe { &mut *self.encoding_context.get() }
+    }
+
+    // SAFETY: This method takes an immutable reference to `self` and then modifies the
+    //         EncodingContext's bump allocator via `UnsafeCell`. This should only be called from
+    //         `between_top_level_values`, and the caller must confirm that nothing else holds a
+    //         reference to any structures within `EncodingContext`.
+    unsafe fn reset_bump_allocator(&self) {
+        let context: &mut EncodingContext = &mut *self.encoding_context.get();
+        context.allocator.reset();
     }
 
-    pub fn pending_symtab_changes(&self) -> &PendingLst {
+    pub fn pending_lst(&self) -> &PendingLst {
         // If the user is able to call this method, the PendingLst is not being modified and it's
         // safe to immutably reference.
         unsafe { &*self.pending_lst.get() }
     }
 
+    pub fn pending_lst_mut(&mut self) -> &mut PendingLst {
+        // SAFETY: If the caller has a `&mut` reference to `self`, it is the only mutable reference
+        //         that can modify `self.pending_lst`.
+        unsafe { &mut *self.pending_lst.get() }
+    }
```

Comment:
🗺️ These methods where the mutability of the `self` reference (`&self` or `&mut self`) matches the mutability of the returned reference are only unsafe because the type being accessed is an `UnsafeCell`